src/Qowaiv.Validation.Xml/Extensions/System.Xml.Linq.XObject.cs

```diff
@@ -0,0 +1,94 @@
+        => document.Validate(stream.AsXmlSchema());
+
+    /// <summary>Validates the document using a <see cref="Stream"/> representing an <see cref="XmlSchema"/>.</summary>
+    [Pure]
+    public static Result<XDocument> Validate(this XDocument document, XmlSchema schema)
+        => document.Validate(new XmlSchemaSet().Append(schema));
+
+    /// <summary>Validates the document using an <see cref="XmlSchemaSet"/>.</summary>
+    [Pure]
+    public static Result<XDocument> Validate(this XDocument document, XmlSchemaSet schemas)
+    {
+        Guard.NotNull(document, nameof(document));
+        Guard.NotNull(schemas, nameof(schemas));
+
+        var handler = new XmlValidationHandler();
+        document.Validate(schemas, handler.Validate);
+        return Result.For(document, handler.Messages);
+    }
+
+    /// <summary>Get the absolute XPath of the attribute.</summary>
+    /// <remarks>
+    /// e.g.: /people/person[6]/name[1]/last[1]/@prefx
+    /// </remarks>
+    [Pure]
+    internal static string AbsoluteXPath(this XAttribute attribute)
+    {
+        Guard.NotNull(attribute, nameof(attribute));
+        return $"{attribute.Parent?.AbsoluteXPath()}/@{attribute.Name.LocalName}";
+    }
+
+    /// <summary>Get the absolute XPath of the element.</summary>
+    /// <remarks>
+    /// e.g.: /people/person[6]/name[1]/last[1]
+    /// </remarks>
+    [Pure]
+    internal static string AbsoluteXPath(this XElement element)
+    {
+        Guard.NotNull(element, nameof(element));
+        return string.Concat(element
+            .AncestorsAndSelf().Select(RelativePath).Reverse());
+    }
+
+    [Pure]
+    private static string RelativePath(this XElement element)
+    {
+        int index = element.IndexPosition();
+        string name = element.Name.LocalName;
+
+        if (element.AncestorsAndSelf()
+            .SelectMany(x => x.Attributes())
+            .FirstOrDefault(x => x.IsNamespaceDeclaration && x.Value == element.Name.NamespaceName) is { } alias
+            && !string.IsNullOrEmpty(alias.Name.NamespaceName)
+            && !string.IsNullOrEmpty(alias.Name.LocalName))
+        {
+            name = $"{alias.Name.LocalName}:{name}";
+        }
+        return index == -1
+            ? $"/{name}"
+            : $"/{name}[{index}]";
+    }
+
+    /// <summary>
+    /// Get the index of the given XElement relative to its
+    /// siblings with identical names. If the given element is
+    /// the root, -1 is returned.
+    /// </summary>
+    /// <param name="element">
+    /// The element to get the index of.
+    /// </param>
+    [Pure]
+    private static int IndexPosition(this XElement element)
+    {
+        if (element.Parent is { })
+        {
+            var index = 0;
+
+            foreach (var sibling in element.Parent.Elements(element.Name))
+            {
+                index++;
+                if (sibling == element) return index;
```

Comment:
That should not be the case. It might be possible to add a reference of an existing node to somewhere else in the document, but there is no entry point in this API that allows to manipulate the created XDocument.